tableaudocumentapi/datasource.py

```diff
@@ -258,3 +259,38 @@ def _get_metadata_objects(self):
     def _get_column_objects(self):
         return [_column_object_from_column_xml(self._datasourceTree, xml)
                 for xml in self._datasourceTree.findall('.//column')]
+
+    ###########
+    # folders
+    ###########
+    @property
+    def folders(self):
+        if not self._folders:
+            self._refresh_folders()
+        return self._folders
+
+    def _refresh_folders(self):
+        folders = Folder.all_folders_from_datasource(self)
+        self._folders = {f.name: f for f in folders}
+
+    def add_folder(self, name, role):
+        """ Adds a new, empty folder to the datasource and returns it.
+
+        Will fail if another folder of the same name already exists,
+        or if no valid role (dimensions or measures) is provided.
+        """
+
+        if name in self.folders.keys():
+            raise ValueError('Folder names must be unique')
+
+        folder = Folder.from_name_and_role(self, name, role)
```

Comment:
Got it!

Thanks for the explanation. 

To me, passing just `self` isn't a preferred way of doing things because it couples the classes really tightly. Renaming the variable is better but still tightly coupled.

However, I just got back from a New Years Eve trip to SF and my brain is slow, so I don't have an alternative suggestion yet :)